### PR TITLE
Cleaning up summary and doc metadata for ephemeral containers

### PR DESCRIPTION
**Pull request by pradeepvairamani**
## Description

- This PR deletes the summary related artifacts and the document metadata for ephemeral containers at the end of the session.
- Added recursive rm support
- Also, made a change to throw a 404 instead of 500 when commits are not found for ephemeral containers
- A few telemetry improvements

### CODE REVIEW

File: server/gitrest/packages/gitrest-base/src/utils/isomorphicgitManager.ts [hunk 1]
```diff
@@ -88,6 +89,9 @@ export class IsomorphicGitRepositoryManager extends RepositoryManagerBase {
 				},
 				err,
 			);
+			if (this.isEphemeralContainer && err?.code === "NotFoundError") {
```
Comment (manishgargd) [inline server/gitrest/packages/gitrest-base/src/utils/isomorphicgitManager.ts:92]: If err?.code is 404, we can throw 404 even if it is not an EphemeralContainer.

Comment (pradeepvairamani) [reply server/gitrest/packages/gitrest-base/src/utils/isomorphicgitManager.ts:92]: That is a good question. With 404, we are telling the clients to give up on the document and create a new one. This is totally fine for ephemeral containers but it should ideally not happen in the durable containers. I think it comes down to whether we want the clients to retry or give up for durable containers. I think @hedasilv and @znewton might be in a better suited to make that call.

Comment (znewton) [reply server/gitrest/packages/gitrest-base/src/utils/isomorphicgitManager.ts:92]: @pradeepvairamani is correct about the impact of that type of change. There needs to be a larger effort to cleanup and improve Gitrest's error codes, and I don't think we should increase the scope of this PR with that work.

There are some parts of R11s (including the driver) that rely on specific error codes to operate correctly. For example, `getOrCreateRepository` from services-client explicitly requires a `400`, instead of the more logical and correct `404`, to create a repo when get repo fails. Another example is `GitManager.getRef`, which returns `null` on 400 or 404, or throws otherwise. We need to be careful when changing that error code so that we don't break what was previously working with a 400.

TL;DR: Changing Gitrest error codes should receive a sizeable amount of attention and validation, and I don't think this is the right time/PR to do that.

File: server/routerlicious/packages/lambdas/src/deli/lambdaFactory.ts [hunk 2]
```diff
@@ -225,6 +225,22 @@ export class DeliLambdaFactory
 					closeType === LambdaCloseType.ActivityTimeout ||
 					closeType === LambdaCloseType.Error
 				) {
+					if (document?.isEphemeralContainer) {
+						// Call to historian to delete summaries
+						await gitManager.deleteSummary(false);
```
Comment (znewton) [inline server/routerlicious/packages/lambdas/src/deli/lambdaFactory.ts:230]: should we `requestWithRetry/runWithRetry` on these 2 deletion operations?

File: server/gitrest/packages/gitrest-base/src/utils/redisFs/redis.ts [hunk 3]
```diff
@@ -52,7 +52,7 @@ export class Redis {
 	}
 
 	public async delete(key: string): Promise<boolean> {
-		const result = await this.client.del(this.getKey(key));
+		const result = await this.client.del(key);
```
Comment (manishgargd) [inline server/gitrest/packages/gitrest-base/src/utils/redisFs/redis.ts:55]: Just want to make sure that this change is intentional. Previously we were using thig.getKey(key).

Comment (pradeepvairamani) [reply server/gitrest/packages/gitrest-base/src/utils/redisFs/redis.ts:55]: This is intentional. In the redis Fs, the pattern that we are using is that we have a particular folder path for a doc and we want to delete all the values related to that doc. While we enumerate what we want to delete (like using redis keys*), we get the keys with the prefix. If we getKey again here, it would prefix the key twice. Also, this method is not being used anywhere else, so it is a safe change.

File: server/routerlicious/packages/lambdas/src/deli/lambdaFactory.ts (hunk 2 above)
Comment (pradeepvairamani) [reply server/routerlicious/packages/lambdas/src/deli/lambdaFactory.ts]: good idea..done

File: server/gitrest/packages/gitrest-base/src/utils/redisFs/redis.ts (hunk 3 above)
Comment (znewton) [reply server/gitrest/packages/gitrest-base/src/utils/redisFs/redis.ts:55]: I agree that this is safe, since this Redis class is within `redisFs/`, but for the sake of future maintenance, could you add a quick code comment explaining like you did in this PR comment?

**Review by znewton — APPROVED** [review]

File: server/gitrest/packages/gitrest-base/src/utils/redisFs/redis.ts (hunk 3 above)
Comment (hedasilv) [reply server/gitrest/packages/gitrest-base/src/utils/redisFs/redis.ts:55]: Just to clarify: earlier, we would delete the key in Redis after adding the appropriate prefix.

now that we just call `this.client.del(key)`, will the command succeed? Are we not setting the key with the prefix every time?

I might be failing to understand the intention here. Or maybe, is it that we always call `delete()` after enumerating the keys, which naturally come with a prefix and therefore we don't need to add the prefix again?

File: server/routerlicious/packages/lambdas/src/deli/lambdaFactory.ts [hunk 4]
```diff
@@ -225,6 +227,38 @@ export class DeliLambdaFactory
 					closeType === LambdaCloseType.ActivityTimeout ||
 					closeType === LambdaCloseType.Error
 				) {
+					if (document?.isEphemeralContainer) {
```
Comment (hedasilv) [inline server/routerlicious/packages/lambdas/src/deli/lambdaFactory.ts:230]: Now that Deli will be making HTTP requests to the summary endpoint, we might need to check if there is anything else needed from the point of view of security/compliance. I think we are already running LinkerD within the Deli pods, but if not, we might need to do that

File: server/routerlicious/packages/lambdas/src/deli/lambdaFactory.ts [hunk 5]
```diff
@@ -225,6 +227,38 @@ export class DeliLambdaFactory
...
+
+						// Delete the document metadata
+						await runWithRetry(
+							async () =>
+								this.documentRepository.deleteOne({
+									documentId,
+									tenantId,
+								}),
+							"deleteDocMetadata",
+							3 /* maxRetries */,
```
Comment (hedasilv) [inline server/routerlicious/packages/lambdas/src/deli/lambdaFactory.ts:248]: What if this fails, but the call from line 232 above succeeds? Would it be a problem to still have the document info in the DB, but no more summary data in (redis) storage?

Comment (hedasilv) [reply server/routerlicious/packages/lambdas/src/deli/lambdaFactory.ts:248]: Perhaps we can add a try/catch here with specific error handling and logging for those situations? We do have try/catch in runWithRetry and requestWithRetry but those are more generic.

File: server/routerlicious/packages/lambdas/src/deli/lambdaFactory.ts (hunk 4 above)
Comment (pradeepvairamani) [reply server/routerlicious/packages/lambdas/src/deli/lambdaFactory.ts:230]: We already have HTTP calls that we are making to other microservices like riddler. So, this is not a new change to the that behavior. But just to be sure, I made sure that we are running linkerd in deli

image

File: server/routerlicious/packages/lambdas/src/deli/lambdaFactory.ts (hunk 5 above)
Comment (pradeepvairamani) [reply server/routerlicious/packages/lambdas/src/deli/lambdaFactory.ts:248]: That would not be a problem. The only difference is that it will fail at a different point where it will not be able to fetch the summary. It is to handle this scenario that I had to return a 404 instead of a 500 for ephemeral containers, so that the clients can react accordingly.

I think these calls are not important enough to track individually. Since the session is already winding down, these failing should not affect the session. It is just that the session cleanup will happen at a later point when the TTL expires. Also, the error stack for both the calls are quite unique, so it should be sufficient to tell them apart.

File: server/gitrest/packages/gitrest-base/src/utils/redisFs/redis.ts (hunk 3 above)
Comment (pradeepvairamani) [reply server/gitrest/packages/gitrest-base/src/utils/redisFs/redis.ts]: Modified this to make the caller pass in the param if we want to ignore the prefixing. Hope this makes it more clear.

File: server/gitrest/packages/gitrest-base/src/utils/redisFs/redis.ts [hunk 6]
```diff
@@ -51,8 +51,9 @@ export class Redis {
 		}
 	}
 
-	public async delete(key: string): Promise<boolean> {
-		const result = await this.client.del(key);
+	public async delete(key: string, appendPrefixToKey = true): Promise<boolean> {
+		const keyToDelete = appendPrefixToKey ? this.getKey(key) : key;
```
Comment (hedasilv) [inline server/gitrest/packages/gitrest-base/src/utils/redisFs/redis.ts:55]: Maybe adding the comment @znewton suggested about why we can sometimes not add prefix would also be helpful for context in the future when we revisit the code

**Review by hedasilv — APPROVED** [review]

File: server/routerlicious/packages/lambdas/src/deli/lambdaFactory.ts (hunk 4 above)
Comment (hedasilv) [reply server/routerlicious/packages/lambdas/src/deli/lambdaFactory.ts:230]: Thanks for confirming! 😄

File: server/routerlicious/packages/lambdas/src/deli/lambdaFactory.ts (hunk 5 above)
Comment (hedasilv) [reply server/routerlicious/packages/lambdas/src/deli/lambdaFactory.ts:248]: Sounds good. I agree we should be good since all the data will have a TTL.

**Review by hedasilv — APPROVED** [review]
